[Intel Mkl] Parallel BiasAddGrad op with eigen intra thread pool

wenxizhu opened the pull request:
The BiasAddGrad op is running on single thread, which badly influences the training performance of corresponding models.

We provide a optimized parallel implementation of BiasAddGrad op.

The following is the time cost for original and optimized BiasAddGrad op of some corresponding models. All run on skx-8180 single socket:
YoloV2 (batch size 16) : 263.76ms -> 12.06ms
Trans-LT (batch size 1024) : 68.61ms -> 3.09ms
Inception-ResV2 (batch size 64) : 334.96ms -> 73.18ms
NCF (batch size 1024) : 1.34ms -> 0.243ms
NCF-mlperf (batch size 1024) : 0.867ms -> 0.341ms
MaskRCNN (batch size 1) : 370.57ms -> 15.23ms
DCGAN (batch size 32) : 10.46ms -> 0.512ms

JoursBleu committed:
The BiasAddGrad op is running on single thread, which badly
influences the training performance of corresponding models.

We provide a optimized parallel implementation of BiasAddGrad op.

Change-Id: I3a8da878eea67a4903a3b68302c6c86c3e536025

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -18,13 +18,15 @@ limitations under the License.
 #include "tensorflow/core/util/tensor_format.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+
+#include "tensorflow/core/util/work_sharder.h"
```
Could you please group all tensorflow includes together.

wenxizhu replied:
Done.

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
  private:
   TensorFormat data_format_;
+
+  // Modified for performance tune :: new funcs
```
I think you can strip all "Modified for performance" comments, they are not really helpful for the reader.

wenxizhu replied:
Removed.

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
+  inline void InplaceVecAdd(T* X, const T* Y, const int64 length) {
+    //#pragma simd
+    for (int64 i = 0; i < length; i++) {
+      X[i] = X[i] + Y[i];
```
Do you know if this pragma works with clang? This loop can be expressed as simple Eigen expression, something like:

```
    using X = Eigen::Map<Eigen::Array<T, Dynamic, 1>>;
    using Y = Eigen::Map<const Eigen::Array<T, Dynamic, 1>>
    X(x, length) += Y(y, length);
```

I probably messed up with some constructor parameters, but similar idea is used in https://bitbucket.org/eigen/eigen/src/9c300336de9a096c7c4c3b230a6ae9f49fa17aea/unsupported/Eigen/CXX11/src/Tensor/TensorBlock.h#lines-494:506

This will be vectorized/unrolled

wenxizhu replied:
Ok, repalced them with eigen apis. And I also figured there is no need to keep the `InplaceVecAdd()` and `VecSumReduce()` fuctions. I inlined them into the lambda function.

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
+
+  // Modified for performance tune :: new funcs
+  // Apply X[0:length-1] = X[0:length-1] + Y[0:length-1];
+  inline void InplaceVecAdd(T* X, const T* Y, const int64 length) {
```
function arg names must be 'x' and 'y'

wenxizhu replied:
This function has been removed, so that's not a problem now.

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
+    }
+  }
+  // Return sum(X[0:length-1])
+  inline T VecSumReduce(const T* X, const int64 length) {
```
a. method arg name
b. it also could be `return Eigen::Map<...>(x, length).sum();`

I didn't find any other simd pragmas in Tensorflow codebase, so my assumption it's not supported by all compilers.

wenxizhu replied:
This function has been removed, so that's not a problem now.

ezhulenev reviewed (commented):
*EDIT:* I prepared this benchmark myself, should be in GitHub soon.

----
*Original:*
Could you also add bias_op_test.cc with micro benchmarks, you can use https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/sparse_xent_op_test.cc as an example. I will be able to use them to compare improvements with clang toolchain.

ezhulenev commented on tensorflow/core/kernels/bias_op.cc:
```diff
@@ -268,43 +269,184 @@ class BiasGradOp : public OpKernel {
-                .template cast<typename AccumulatorType<T>::type>()
-                .reshape(two_dims)
-                .sum(reduction_axis)
-                .template cast<T>();
```
I think all performance problems with this grad were fixed in https://bitbucket.org/eigen/eigen/commits/948b43eaf2c5c35c988308d120a8a24781c63de6, what version of TF you used in benchmarks?

wenxizhu replied:
According to my measurement, I believe my implementation is still faster than this commit you mentioned. So could I keep using my implementation.

ezhulenev commented:
I did run benchmarks from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/bias_op_test.cc and got pretty good results already, feel free to add more benchmarks, probably you want to test NCHW as well.

```name                                              old time/op             new time/op             delta
BM_BiasAddGradNHWC_32_32_32_128_cpu               4.84ms ± 1%             2.30ms ± 1%  -52.46%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_32_32_256_cpu               9.66ms ± 2%             4.15ms ± 1%  -57.08%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_32_32_512_cpu               80.1ms ±13%             18.4ms ± 4%  -76.97%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_32_32_1024_cpu               242ms ± 2%               41ms ±17%  -83.17%          (p=0.000 n=9+9)
BM_BiasAddGradNHWC_32_64_64_128_cpu               58.3ms ± 3%             18.5ms ± 5%  -68.16%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_64_64_256_cpu                177ms ± 5%               44ms ±12%  -75.48%         (p=0.000 n=9+10)
BM_BiasAddGradNHWC_32_64_64_512_cpu                452ms ± 6%               92ms ± 8%  -79.75%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_64_64_1024_cpu               1.19s ± 5%              0.23s ± 7%  -80.93%        (p=0.000 n=10+10)

name                                              old allocs/op           new allocs/op           delta
BM_BiasAddGradNHWC_32_32_32_128_cpu                  106 ± 0%                 90 ± 0%  -15.09%         (p=0.000 n=10+9)
BM_BiasAddGradNHWC_32_32_32_256_cpu                  170 ± 0%                 90 ± 0%  -47.06%         (p=0.000 n=10+9)
BM_BiasAddGradNHWC_32_32_32_512_cpu                  170 ± 0%                 90 ± 1%  -47.35%         (p=0.000 n=8+10)
BM_BiasAddGradNHWC_32_32_32_1024_cpu                 170 ± 0%                 90 ± 1%  -47.33%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_64_64_128_cpu                  107 ± 1%                 90 ± 0%  -15.57%         (p=0.000 n=10+9)
BM_BiasAddGradNHWC_32_64_64_256_cpu                  171 ± 0%                 90 ± 0%  -47.37%          (p=0.001 n=8+9)
BM_BiasAddGradNHWC_32_64_64_512_cpu                  170 ± 0%                 90 ± 1%  -47.33%        (p=0.000 n=10+10)
BM_BiasAddGradNHWC_32_64_64_1024_cpu                 170 ± 0%                 90 ± 1%  -47.48%        (p=0.000 n=10+10)
```

JoursBleu committed:
Codes normalization:
  1. Group all tensorflow includes together.
  2. Delete unnecessary comments.
  3. Using eigen ops instead.

Change-Id: I67edcd71cb4feeaf8ab0c1820d2c011e3409344d

ezhulenev commented:
Based on your PR I've submitted very similar change in https://github.com/tensorflow/tensorflow/commit/d5f6595171bb21e19da1c7dc75d02f3df85cf12b, apparently this "reduce all outer dimension" is pretty common in other gradient kernels (e.g. in FusedBatchNorm https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/fused_batch_norm_op.cc#L215). Can you please move the code for NCHW case to `redux_functors.h`, and call it similar way. Not sure though what's a good name for it.

Currently it takes Tensor as an input, but it could be any Eigen expression, I'll work on it after this PR will be merged.

JoursBleu commented:
Hi @ezhulenev ,
I have read the codes in ReduceOuterDimensions.
So, you want me to add a functor for situations like:
    (32, 11, 256, 256) -> (11);
or maybe more general like:
    (D1, D2, ... , DN) -> (DM) (where M belong to set [1,N])
and then ref the new functor to do the reduce?

ezhulenev commented:
Yeah, that’s the idea. I think you can skip the general case, and just add
support for (d1, ... , dn) -> dm where m is in [2, n-1] range. Reducing
inner most dimensions are already fast in Eigen, I added optimized case for
reducing outer dimensions. We just need one more optimized version for
reducing all-except “middle” dimension.
On Mon, Mar 11, 2019 at 11:57 PM Letian <notifications@github.com> wrote:

> Hi @ezhulenev <https://github.com/ezhulenev> ,
> I have read the codes in ReduceOuterDimensions.
> So, you want me to add a functor for situations like:
> (32, 11, 256, 256) -> (11);
> or maybe more general like:
> (D1, D2, ... , DN) -> (DM) (where M belong to set [1,N])
> and then ref the new functor to do the reduce?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/26426#issuecomment-471876004>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHrajX_gC0cA2ojKyuJ4NRtE2P_La-zks5vV0_ugaJpZM4bifIc>
> .
>

JoursBleu commented:
@ezhulenev  OK

JoursBleu commented:
@ezhulenev 
Could you please provide the command to run the test in bias_op_test.cc?

ezhulenev commented:
`bazel run .... -- --benchmarks=all` I think something like that, but I'm not 100% sure, it might be some difference between internal/external tooling. Teng Lu (@Zantares) from Intel made it work in https://github.com/tensorflow/tensorflow/pull/26424#issuecomment-472965743, I guess he definitely has the answer :)

JoursBleu committed:
Merge master in

JoursBleu committed:
1. Add ReduceMiddleDimensions to support middle dim
   reduce in tensorflow/core/kernels/redux_functor.h
2. Opt ReduceOuterDimensions for large inner dim.
3. Rewrite NCHW BiasAddGradOp with ReduceOuterDimensions.

JoursBleu committed:
1. Add ReduceMiddleDimensions to support middle dim
   reduce in tensorflow/core/kernels/redux_functor.h
2. Opt ReduceOuterDimensions for large inner dim.
3. Rewrite NCHW BiasAddGradOp with ReduceOuterDimensions.

ezhulenev commented:
This fails internally with a error:

```

InvalidArgumentError: var and grad do not have the same shape[1] [3,1]
	 [[node adagrad/slot_11/update_matmul3_b/ApplyAdagrad (defined at {INTERNAL}/training_graph_generator.py:1749) ]]

Errors may have originated from an input operation.
Input Source operations connected to node adagrad/slot_11/update_matmul3_b/ApplyAdagrad:
 gradients/matmul3/BiasAdd_grad/BiasAddGrad (defined at {INTERNAL}//training_graph_generator.py:1306)	
 matmul3_b (defined at {INTERNAL})

```

didn't have time to debug it yet

JoursBleu commented:
@ezhulenev 
OK, how can I reproduce the error? I can not find the file "training_graph_generator.py" anywhere...

JoursBleu commented:
@ezhulenev is there anything I can do?

ezhulenev commented:
I'll try to prepare reproducible test, right now it's a part of a large model and it's hard to tell whats going on.

JoursBleu commented:
@ezhulenev Thx. Just tell me if there is anything to do. We need this pr merged ASAP, cause several model is depending on this one.

ezhulenev commented on tensorflow/core/kernels/redux_functor.h:
```diff
@@ -16,38 +16,208 @@ limitations under the License.
+    for (int i = num_dims - num_reduce_dims; i < num_dims; ++i)
+      inner_dim *= input_dims[i];
+
+    if (1 == inner_dim) {
```
Isn't it a bug? '32x32x1' should be reduced to scalar?

JoursBleu replied:
Exactly! It should be "outer_dim" here.
Thank you very much for pointing out the bug!
Fixed.

ezhulenev replied:
`tensorflow/python/keras:convolutional_test`

`ValueError: Shapes (3,) and (1, 1, 1, 1, 3) are incompatible`

JoursBleu replied:
fixed

JoursBleu committed:
Fix logical bug in front of ReduceOuterDimensions.

JoursBleu committed:
Merge branch 'master' into letian/biasaddgrad_opt

JoursBleu committed:
Fix shape of result(input when nothing to do) unmatch output error.
